bson/string_codec.go

```diff
@@ -50,14 +51,16 @@ func (sc *stringCodec) decodeType(dc DecodeContext, vr ValueReader, t reflect.Ty
 			return emptyValue, err
 		}
 	case TypeObjectID:
-		oid, err := vr.ReadObjectID()
-		if err != nil {
-			return emptyValue, err
-		}
 		if dc.objectIDAsHexString {
+			oid, err := vr.ReadObjectID()
+			if err != nil {
+				return emptyValue, err
+			}
 			str = oid.Hex()
 		} else {
-			return emptyValue, fmt.Errorf("decoding an object ID to a non-hexadecimal string representation is not supported")
+			msg := "decoding an object ID into a plain string is not supported " +
+				"(set Decoder.ObjectIDAsHexString to enable decoding as a hexadecimal representation)"
```

Comment:
Optional: Use a `const` and slightly clearer wording.

```suggestion
			const msg = "decoding an object ID into a string is not supported by default" +
				"(set Decoder.ObjectIDAsHexString to enable decoding as a hexadecimal string)"
```